### ISSUE TITLE
CLI does not upload to existing dataset

### ISSUE THREAD

**Issue by jrudoler** [OpenNeuroOrg/openneuro#2750]
**Describe the bug**
I have a very large dataset (8.2 TB) that I was unable to upload in one go. Perhaps I will try again and create a separate issue, but for now I'm trying partial uploads to make this more manageable. However, successive uploads are not working either. I uploaded 35 subjects, then another 30 (which seemed to work in the CLI but did not show up online), and now uploading another 30 does not work at all. 

Perhaps related: The subject directories shown in the file list do not match the metadata on the web UI.  I see only 35 subjects (from an earlier upload) but the website claims the # of participants in my dataset is 65. 

**Expected behavior**
According to the docs, new files can be added to an existing dataset to accomplish successive partial uploads of a large dataset. I expect the CLI to validate the local copy, check it against the remote copy to identify new and modified files, and then initiate an upload of those new files.  

**Screenshots**
Here is a screenshot of the CLI. After validating the local copy and checking remote, it prompts for upload approval and then immediately returns without showing the expected progress bar or throwing any errors. This apparently has no effect on the server. 
<img width="934" alt="Screenshot 2023-01-12 at 10 10 20 AM" src="https://user-images.githubusercontent.com/40604402/212103597-9e473ae3-0d8e-416a-a4a8-e6244847d91a.png">

This upload is strictly adding data, not modifying existing files. I noticed that the CLI's assessment of how many files had been added or changed did not match what I saw on the dataset's snapshot:
<img width="1055" alt="Screenshot 2023-01-12 at 9 59 51 AM" src="https://user-images.githubusercontent.com/40604402/212103659-a6c598dd-c100-48a4-8a9a-a917a62602cb.png">
I expected to see, roughly: total local size = total remote size + upload size. This is not the case.

**Additional context**
OS: CentOS Linux 7
OpenNeuro: 4.14.1
Node.js: v16.17.0
npm: 8.19.2

**Comment by nellh** [comment]
With OpenNeuro 4.13.0, the minimum required version of Node.js is Node 18. Can you try with Node 18?

You could also try @openneuro/cli@4.12.1 which is the most recent version to support Node 16. It will warn about being out of date but at the moment it's still API compatible with the OpenNeuro service.

**Comment by jrudoler** [comment]
Using Node18: now it at least tries to upload and produces errors: these request timeouts are the same error I got when trying to upload the whole dataset originally.
<img width="2672" alt="Screenshot 2023-01-12 at 12 22 53 PM" src="https://user-images.githubusercontent.com/40604402/212136327-0bc165c2-f1bc-4fa3-a43f-9f7fce63ea30.png">

**Comment by nellh** [comment]
> Using Node18: now it at least tries to upload and produces errors: these request timeouts are the same error I got when trying to upload the whole dataset originally. <img alt="Screenshot 2023-01-12 at 12 22 53 PM" width="2000" src="https://user-images.githubusercontent.com/40604402/212136327-0bc165c2-f1bc-4fa3-a43f-9f7fce63ea30.png">

Thanks, that's very helpful for narrowing this down. Looks like the same issue as #2748, as a workaround you can try the @openneuro/cli@4.12.1 release, that should avoid this problem. With Node 18, you'll need to run the 4.12.1 release with `NODE_OPTIONS=--no-experimental-fetch` set.

**Comment by jrudoler** [comment]
I will give it a shot... I am already running `export NODE_OPTIONS=--max_old_space_size=80000` to avoid the out-of-memory killer in nodejs. I'm running the upload on an HPC node with 80GB allocated (I know this is overkill, but wanted to be sure memory was not the issue).

**Comment by jrudoler** [comment]
This results in the same issue as before (plus a million version warnings) - cli returns with no error messages, and no progress bar or evidence of attempting the upload. 

<img width="2672" alt="Screenshot 2023-01-12 at 12 57 22 PM" src="https://user-images.githubusercontent.com/40604402/212143146-51ac7384-28ca-4dd6-93a4-b8e5078f754a.png">

**Comment by nellh** [comment]
Thanks, a fix for this is in review for #2752 and I'll let you know when it is released in OpenNeuro 4.14.2

Do things work with a more sensible max_old_space_size (2048 maybe)? Raising it is sometimes required for validation of larger datasets but I'd be surprised if it doesn't work with something like 2-4GB and that may be worth looking into separately.

**Comment by jrudoler** [comment]
Thanks! Keep me posted.

When I checked the default value (before specifying the parameter explicitly) it was 4096 (or something close to that) and the job was killed after indeed using up about 4GB (slurm tracks the max memory used by an interactive job).

I documented the output back when that was happening, and before I found the max_old_space_size workaround:
<img width="1021" alt="Screenshot 2023-01-12 at 1 43 13 PM" src="https://user-images.githubusercontent.com/40604402/212152694-2935163e-d0c0-460a-bdf6-25e9d15a0db2.png">

**Comment by nellh** [comment]
OpenNeuro 4.14.2 is now on npm and has several changes that should address this. Please give it a try and let me know if you are still having this issue.

**Comment by jrudoler** [comment]
@nellh Is NODE_OPTIONS=--no-experimental-fetch still necessary with this version?

**Comment by nellh** [comment]
No, disabling Node.js fetch is only required for 4.12.1 or earlier with Node 18 or later.

**Comment by jrudoler** [comment]
I still get the same warning, though I have not gotten request timeouts yet:
<img width="2054" alt="Screenshot 2023-01-12 at 3 23 33 PM" src="https://user-images.githubusercontent.com/40604402/212173426-3889b417-7db9-408c-92a4-bfcca77f53ea.png">

**Comment by jrudoler** [comment]
This did not resolve the issue. Still getting the same errors!

**Comment by nellh** [comment]
> This did not resolve the issue. Still getting the same errors!

Thanks for your help debugging so far. Could you try another possible fix? I've tagged a prerelease @openneuro/cli@4.13.3-alpha.1 with another possible solution.

**Comment by jrudoler** [comment]
Glad to help sort things out - appreciate the quick response time!

Using @openneuro/cli@4.14.3-alpha.1, the upload still did not succeed. Slurm reported that it ran out of the 80GB of allocated memory. If this is not supposed to be a memory-intensive process, then something is amiss. On the bright side, no timeouts or fetch errors. 

<img width="969" alt="Screenshot 2023-01-12 at 7 12 14 PM" src="https://user-images.githubusercontent.com/40604402/212207762-a3ba8808-dc4a-44fa-b9f1-17c7b14a7ebf.png">

**Comment by nellh** [comment]
I think this is at least a different issue, so that's good. I think I found another issue with how we buffered files on retries which lead to holding onto part of multiple files, which could consume a lot of memory if the set of files being uploaded is made up of mostly large files. There should be an improvement in 4.14.3-alpha.5 if you could try that.

For future reference the original problem here causing some requests to fail multiple times seems related to. https://github.com/nodejs/undici/issues/1248

**Comment by jrudoler** [comment]
Great news! Looks like it worked with 4.14.3-alpha.5! 🎉🎉 Thanks so much for your help!

For reference, this upload of ~614 GB used a maximum of ~2.3 GB of memory. With such a manageable memory load, I'll see if it works to upload the remainder of the 8 TB instead of doing this incrementally. I'll report back if it works / max memory usage so we can see how this scales. 

So quick recap:
node: v18.12.1
openneuro: 4.14.3-alpha.5
allocated 80GB of memory (massive overkill) and ran `export NODE_OPTIONS=--max_old_space_size=80000`

**Comment by jrudoler** [comment]
Uploading the full dataset has failed a couple of times, oddly without any error messages. Most recently it reached 95% completion before apparently silently aborting (after more than 24 hours).
<img width="2672" alt="Screenshot 2023-01-16 at 6 08 51 PM" src="https://user-images.githubusercontent.com/40604402/212778370-7a5f92d7-3245-4acd-9e1c-9d39f84009f3.png">

**Comment by nellh** [comment]
> Great news! Looks like it worked with 4.14.3-alpha.5! tadatada Thanks so much for your help!
> 
> For reference, this upload of ~614 GB used a maximum of ~2.3 GB of memory. With such a manageable memory load, I'll see if it works to upload the remainder of the 8 TB instead of doing this incrementally. I'll report back if it works / max memory usage so we can see how this scales.
> 
> So quick recap: node: v18.12.1 openneuro: 4.14.3-alpha.5 allocated 80GB of memory (massive overkill) and ran `export NODE_OPTIONS=--max_old_space_size=80000`

Great to hear this is working with more reasonable memory usage now.

> Uploading the full dataset has failed a couple of times, oddly without any error messages. Most recently it reached 95% completion before apparently silently aborting (after more than 24 hours). <img alt="Screenshot 2023-01-16 at 6 08 51 PM" width="2000" src="https://user-images.githubusercontent.com/40604402/212778370-7a5f92d7-3245-4acd-9e1c-9d39f84009f3.png">

This is definitely a stress test of OpenNeuro, this dataset is about 4x the largest we've accept so far. It looks like the server process ran out memory building up the git-annex commit. I gave it some additional memory and it seems to be working now (though it will take a while to checksum), once it's finished I'll post again and you should be able to resume the upload to add anything that is still missing.

**Comment by jrudoler** [comment]
Yeah - it's a pretty massive dataset. Processing it all and converting to BIDS was a stress test for our HPC too. 
Oh, great - I was worried that that 95% upload was lost. On previous tries to resume aborted uploads it has started from scratch, and the progress was not reflected on the web page... I assumed that this was to prevent incomplete uploads that are non-compliant with BIDS format because of missing files. Is there a good way for me to tell what files have been uploaded (and should not be repeated on subsequent upload)?

**Comment by nellh** [comment]
> Yeah - it's a pretty massive dataset. Processing it all and converting to BIDS was a stress test for our HPC too. Oh, great - I was worried that that 95% upload was lost. On previous tries to resume aborted uploads it has started from scratch, and the progress was not reflected on the web page... I assumed that this was to prevent incomplete uploads that are non-compliant with BIDS format because of missing files. Is there a good way for me to tell what files have been uploaded (and should not be repeated on subsequent upload)?

Once this checksum process finishes, it will generated a commit with the files and they will appear in the draft. Any uploads started after this will skip files that are present on the OpenNeuro side unless the size of the file differs.

**Comment by nellh** [comment]
Just an update, the checksum is still running but about 80% complete now.

**Comment by nellh** [comment]
It's completed now, you should see the files that were successfully uploaded in the dataset draft and a new upload should pick up from there.

**Comment by jrudoler** [comment]
Awesome! I see all the new files, but I'm noticing that the automated metadata (like subject/session count) is not updating - perhaps some process on the server is not working with such a large dataset. Notably, server-side BIDS validation still shows status "Validation Pending". It's also still pending for revision 1.1.0, which was a subset uploaded a few days ago. I'm assuming this validation is supposed to only take a few minutes, like it does locally before upload. 
<img width="436" alt="Screenshot 2023-01-17 at 11 28 27 PM" src="https://user-images.githubusercontent.com/40604402/213085095-0e693996-a783-485d-b11e-0a64f6a04da7.png">

**Comment by jrudoler** [comment]
> Uploading the full dataset has failed a couple of times, oddly without any error messages. Most recently it reached 95% completion before apparently silently aborting (after more than 24 hours).
> <img alt="Screenshot 2023-01-16 at 6 08 51 PM" width="2000" src="https://user-images.githubusercontent.com/40604402/212778370-7a5f92d7-3245-4acd-9e1c-9d39f84009f3.png">

By the way, this upload which reached 95% used a max of 7 GB. Worth noting that this is well above node's default allowed value. Perhaps worth including the NODE_OPTIONS fix in the CLI documentation (though it's definitely an edge case).

**Comment by jrudoler** [comment]
I get the following error when I try to upload the remainder of the dataset:
<img width="1552" alt="Screenshot 2023-01-17 at 11 38 18 PM" src="https://user-images.githubusercontent.com/40604402/213197291-e4f093ac-891d-472d-a24a-eaea1f04e2bb.png">

I have tried multiple times, since it suggests it is temporary - the first drew this error, the second failed without any output (not sure this is related but when I quit the interactive job slurm reported exit code 130, which is critical failure for the `ls` command), and the third produced the same "temporary" error again.

**Comment by nellh** [comment]
Thanks, it actually tries to run the validator on the draft head and cancels the request to upload when it runs out of heap memory during validation, I will raise the memory limit and retest.

**Comment by nellh** [comment]
I've deployed the update to fix this validation hang, I think you should be able to try the upload now. I was able to at least start one (not adding any files but running the check for remote files). It does take a while to run so it may appear to hang for a while.

**Comment by jrudoler** [comment]
Great, it uploaded successfully! I upgraded the CLI to 4.15.0 first. And it looks like the validator and metadata are sorted out too!

A few questions before closing this issue -
1. What happens to dataset metadata that you fill in manually on OpenNeuro.org on the snapshot page (which I think all correspond to dataset_description.json), if you subsequently upload new files (with a different json file, less the data you filled in manually online). Does it merge the changes, or are they completely overwritten by the version you upload? If it does not merge, is there an easy way to pull those metadata changes? As last resort I could always manually download the json file and replace my local copy before downloading.
2. I anticipate needing to make some updates that unfortunately impact EEG file headers, and therefore will require re-uploading some of the EEG files. Am I right to assume that I should hold off on creating a version/snapshot until I make those changes, because that would make a many-TB copy? Want to be conscious of the size of this dataset and not abuse the OpenNeuro servers. Would it be best to create a new, separate dataset with a new dataset id and delete the old one to avoid duplication? In general, if I update metadata and publish a new snapshot (say a new paper is published based on the dataset), should I be worried about duplication? Hoping that it's not a big deal and only the diffs are stored (and therefore it's a trivial amount of added storage).

**Comment by nellh** [comment]
> Great, it uploaded successfully! I upgraded the CLI to 4.15.0 first. And it looks like the validator and metadata are sorted out too!

Glad to hear it!

>  What happens to dataset metadata that you fill in manually on OpenNeuro.org on the snapshot page (which I think all correspond to dataset_description.json), if you subsequently upload new files (with a different json file, less the data you filled in manually online). Does it merge the changes, or are they completely overwritten by the version you upload? If it does not merge, is there an easy way to pull those metadata changes? As last resort I could always manually download the json file and replace my local copy before downloading.

OpenNeuro just replaces the file when a new one is uploaded. You would need to download it to retain the changes made online. You can download the git tree for the dataset with instructions [here](https://docs.openneuro.org/git.html) if you want to make bidirectional changes (you can tell git-annex to import the local copy you already have too to avoid downloading the large files, the repo for this dataset is about 150MB without the annexed objects to give you an idea of the size).

> I anticipate needing to make some updates that unfortunately impact EEG file headers, and therefore will require re-uploading some of the EEG files. Am I right to assume that I should hold off on creating a version/snapshot until I make those changes, because that would make a many-TB copy? Want to be conscious of the size of this dataset and not abuse the OpenNeuro servers. Would it be best to create a new, separate dataset with a new dataset id and delete the old one to avoid duplication? In general, if I update metadata and publish a new snapshot (say a new paper is published based on the dataset), should I be worried about duplication? Hoping that it's not a big deal and only the diffs are stored (and therefore it's a trivial amount of added storage).

For the large files they are kept in git-annex which does not diff binary objects. If they are not included in a snapshot, it is possible to drop the old versions of the binary files, but you'll need to contact us to do that (via GitHub or our support portal). Behind the scenes it is a git squash to merge the commits that added the old and new versions.

**Comment by jrudoler** [comment]
I was able to upload the changes without any problem, and it looks like the metadata have updated after the bids-validator succeeded. 
But, the files still haven't changed after waiting a few days - is another checksum still running on the whole dataset?

**Comment by nellh** [comment]
> I was able to upload the changes without any problem, and it looks like the metadata have updated after the bids-validator succeeded. But, the files still haven't changed after waiting a few days - is another checksum still running on the whole dataset?

Yeah, timeout on the new changes being added to git-annex. I'm rerunning it with no time limit now, I'll update you when it finishes.

**Comment by nellh** [comment]
Okay, those changes should be good now. I've also squashed the commits since 1.1.0 which will save on the storage required.

**Comment by jrudoler** [comment]
Hey @nellh sorry to resurface this, had a delay while trying to iron out some final details before release (specifically participant info and event description files). 
I just uploaded an additional 70GB of changes, which finished uploading yesterday. No changes show up (validator status, metadata, files) on the web interface, though. Mind taking a look and making sure it's not running into time limits again?
Once this is done I'm planning to turn the draft into a version and publish a snapshot. If there's an order in which I should do these things to minimize redundant checksums / waiting time please let me know.

**Comment by nellh** [comment]
> Hey @nellh sorry to resurface this, had a delay while trying to iron out some final details before release (specifically participant info and event description files). I just uploaded an additional 70GB of changes, which finished uploading yesterday. No changes show up (validator status, metadata, files) on the web interface, though. Mind taking a look and making sure it's not running into time limits again? Once this is done I'm planning to turn the draft into a version and publish a snapshot. If there's an order in which I should do these things to minimize redundant checksums / waiting time please let me know.

Apologies for my delay responding, I've been out for a bit. I was able to resolve the issues with this dataset it should be ready for snapshot again. I've also narrowed down the cause here and #2818 is tracking a permanent fix for this problem. Please let me know if you run into issues publishing the dataset.

**Comment by nellh** [comment]
@jrudoler FYI this appears to be exporting as expected, it will take up to a day or so to be fully available for download via GitHub and S3.
